component/kafka/component.go

```diff
@@ -0,0 +1,527 @@
+	if name == "" {
+		errs = append(errs, errors.New("name is required"))
+	}
+
+	if group == "" {
+		errs = append(errs, errors.New("consumer group is required"))
+	}
+
+	if validation.IsStringSliceEmpty(brokers) {
+		errs = append(errs, errors.New("brokers are empty or have an empty value"))
+	}
+
+	if validation.IsStringSliceEmpty(topics) {
+		errs = append(errs, errors.New("topics are empty or have an empty value"))
+	}
+
+	if proc == nil {
+		errs = append(errs, errors.New("work processor is required"))
+	}
+
+	return &Builder{
+		name:         name,
+		group:        group,
+		brokers:      brokers,
+		topics:       topics,
+		proc:         proc,
+		errors:       errs,
+		batchSize:    1,
+		batchTimeout: 100 * time.Millisecond,
+	}
+}
+
+// WithFailureStrategy defines the failure strategy to be used
+// default value is ExitStrategy
+// it will append an error to the builder if the strategy is not one of the pre-defined ones.
+func (cb *Builder) WithFailureStrategy(fs FailStrategy) *Builder {
+	if fs > SkipStrategy || fs < ExitStrategy {
+		cb.errors = append(cb.errors, errors.New("invalid failure strategy provided"))
+	} else {
+		log.Infof(propSetMSG, "failure strategy", cb.name)
+		cb.failStrategy = fs
+	}
+	return cb
+}
+
+// WithRetries specifies the number of retries to be executed per failed processing operation
+// default value is '0'.
+func (cb *Builder) WithRetries(retries uint) *Builder {
+	log.Infof(propSetMSG, "retries", cb.name)
+	cb.retries = retries
+	return cb
+}
+
+// WithRetryWait specifies the duration for the component to wait between retrying processing of messages
+// default value is '0'
+// it will append an error to the builder if the value is smaller than '0'.
+func (cb *Builder) WithRetryWait(retryWait time.Duration) *Builder {
+	if retryWait < 0 {
+		cb.errors = append(cb.errors, errors.New("invalid retry wait provided"))
+	} else {
+		log.Infof(propSetMSG, "retryWait", cb.name)
+		cb.retryWait = retryWait
+	}
+	return cb
+}
+
+// WithBatching specifies the batch size and timeout of the Kafka consumer component.
+// If the size is reached then the batch of messages is processed. Otherwise if the timeout elapses
+// without new messages coming in, the messages in the buffer would get processed as a batch.
+func (cb *Builder) WithBatching(size uint, timeout time.Duration) *Builder {
+	if size == 0 {
+		cb.errors = append(cb.errors, errors.New("zero batch size provided"))
+	} else {
+		log.Infof(propSetMSG, "batchSize", cb.name)
+		cb.batchSize = size
+	}
+
+	if timeout < 0 {
+		cb.errors = append(cb.errors, errors.New("invalid batch timeout provided"))
+	} else {
+		log.Infof(propSetMSG, "batchTimeout", cb.name)
+		cb.batchTimeout = timeout
+	}
+
+	cb.batchTimeout = timeout
+	return cb
+}
+
+// WithSyncCommit specifies that a consumer should commit offsets at the end of processing
+// each message or batch of messages in a blocking synchronous operation.
+func (cb *Builder) WithSyncCommit() *Builder {
```

Comment:
In the SQS package we leave the decision for ACK or NACK to the developer. This way we give them full freedom on what they are doing and actually when they want to do it.

Reply:
The `sync commit` options is different from leaving the decision to the developer in my opinion. Sarama does `MarkMessage` which marks the offset and commits via a clock (by default every one second). The sync commit option here will make sure to commit every time you mark a message.

We can also have a method on the `Batch` interface to `Commit()`. I thought to leave the `MarkMessage` and `SpanSuccess` as private responsibilities of the consumer. The concept of `ACK` suggest a presence of `NACK` which are not very good names when dealing with kafka messages. Do you have any other suggestions that `ACK`? Perhaps `MarkSuccess`? `MarkProcessed`?

Reply:
In Kafka `ACK` is only a lingo of the producer https://www.cloudkarafka.com/blog/2019-09-28-what-does-in-sync-in-apache-kafka-really-mean.html

Just trying to argue that the component should match the terms of the underlying technology.

Reply:
About the naming you are right to use something closer to Kafka. After all it is a Kafka component.
About MarkSuccess we have the following things to decide:
- If autocommit is enabled then a message will be autocommited even if we do not want it. That's why i mentioned to leave the whole Mark thing entirely up to the developer. This way there is only one approach to the whole commit story.
- When processing a batch things are a little bit complicated. In SQS you can mark individual messages or you can mark the whole batch as either ACK or NACK. In Kafka the thing is really complicated because marking and offset marks all previous offsets for the partition as done.
- In SQS is used NACK only for closing the span. How should we deal with this here? Maybe the consumer completes the span after processing internally, this way Nack is not needed and Mark does just Mark.
Some thoughts.

Reply:
Introduced `MarkOffset()` and `Commit()` methods and got ride of `WithSyncCommit()`. Effectively leaving the decision to the developer.